pkg/telnetlib/connection.go

```diff
@@ -0,0 +1,316 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package telnetlib
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"net"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+type connOpts struct {
+	conn        net.Conn
+	fsm         *telnetFSM
+	cmdHandler  CmdHandlerFunc
+	dataHandler DataHandlerFunc
+	serverOpts  map[byte]bool
+	clientOpts  map[byte]bool
+	optCallback func(byte, byte)
+}
+
+type TelnetConn struct {
```

Comment:
It's probably obvious what all of these are for just by their names, but (particularly the *Handler functions) not all of them are.  Can you add some comments as to what each does and where it's used.